turbo/shards/state_cache.go

```diff
@@ -0,0 +1,957 @@
+		if c == 0 {
+			if si.incarnation == i.incarnation {
+				// Code comes before storage items
+				return false
+			}
+			return si.incarnation < i.incarnation
+		}
+		return c < 0
+	default:
+		panic(fmt.Sprintf("unrecognised type of cache item: %T", than))
+	}
+}
+
+func (swi *StorageWriteItem) Less(than btree.Item) bool {
+	switch i := than.(type) {
+	case *AccountWriteItem:
+		c := bytes.Compare(swi.address.Bytes(), i.address.Bytes())
+		if c == 0 {
+			// Account comes before its storage items
+			return false
+		}
+		return c < 0
+	case *StorageWriteItem:
+		c := bytes.Compare(swi.address.Bytes(), i.address.Bytes())
+		if c == 0 {
+			if swi.si.incarnation == i.si.incarnation {
+				return bytes.Compare(swi.location.Bytes(), i.location.Bytes()) < 0
+			}
+			return swi.si.incarnation < i.si.incarnation
+		}
+		return c < 0
+	case *CodeWriteItem:
+		c := bytes.Compare(swi.address.Bytes(), i.address.Bytes())
+		if c == 0 {
+			if swi.si.incarnation == i.ci.incarnation {
+				// Code comes before storage items
+				return false
+			}
+			return swi.si.incarnation < i.ci.incarnation
+		}
+		return c < 0
+	default:
+		panic(fmt.Sprintf("unrecognised type of cache item: %T", than))
+	}
+}
+
+func (swi *StorageWriteItem) GetCacheItem() CacheItem {
+	return swi.si
+}
+
+func (swi *StorageWriteItem) SetCacheItem(item CacheItem) {
+	swi.si = item.(*StorageItem)
+}
+
+func (swi *StorageWriteItem) GetSize() int {
+	return storageWriteItemSize
+}
+
+func (si *StorageItem) GetSequence() int {
+	return si.sequence
+}
+
+func (si *StorageItem) SetSequence(sequence int) {
+	si.sequence = sequence
+}
+
+func (si *StorageItem) GetSize() int {
+	return storageItemSize
+}
+
+func (si *StorageItem) GetQueuePos() int {
+	return si.queuePos
+}
+
+func (si *StorageItem) SetQueuePos(pos int) {
+	si.queuePos = pos
+}
+
+func (si *StorageItem) HasFlag(flag uint16) bool {
+	return si.flags&flag != 0
+}
+
+func (si *StorageItem) SetFlags(flags uint16) {
+	si.flags |= flags
+}
+
+func (si *StorageItem) ClearFlags(flags uint16) {
+	si.flags &^= flags
+}
+
+func (si *StorageItem) String() string {
+	return fmt.Sprintf("StorageItem(addrHash=%x,incarnation=%d,locHash=%x)", si.addrHash, si.incarnation, si.locHash)
+}
+
+func (si *StorageItem) CopyValueFrom(item CacheItem) {
+	otherSi, ok := item.(*StorageItem)
+	if !ok {
+		panic(fmt.Sprintf("expected StorageCacheItem, got %T", item))
+	}
+	si.value.Set(&otherSi.value)
+}
+
+func (ci *CodeItem) Less(than btree.Item) bool {
+	switch i := than.(type) {
+	case *AccountItem:
+		c := bytes.Compare(ci.addrHash.Bytes(), i.addrHash.Bytes())
+		if c == 0 {
+			// Account before its code
+			return false
+		}
+		return c < 0
+	case *StorageItem:
+		c := bytes.Compare(ci.addrHash.Bytes(), i.addrHash.Bytes())
+		if c == 0 {
+			// Code comes before storage items
+			if ci.incarnation == i.incarnation {
+				return true
+			}
+			return ci.incarnation < i.incarnation
+		}
+		return c < 0
```

Comment:
you don't compare storage hash, will it lead to: "2 updates of 1 storage record will produce 2 different records in tree"?

Reply:
This is where `CodeItem` is compared with`StorageItem`, not two `StorageItem`s compared with each others. Therefore, `CodeItem` does not have location hash to compare with. Above, there is a function `Less` for `StorageItem` that handles the case you are thinking of.